clients/application-kafka/src/main/java/org/eclipse/hono/application/client/kafka/impl/KafkaBasedCommandSender.java

```diff
@@ -157,4 +264,123 @@ public KafkaBasedCommandSender(
 
         return props;
     }
+
+    private Handler<DownstreamMessage<KafkaMessageContext>> getCommandResponseHandler(final String tenantId,
+            final Span span) {
+        return message -> {
+            if (message.getCorrelationId() != null) {
+                // if the correlation id of the received response matches with that of the command
+                Optional.ofNullable(pendingCommandResponses.get(tenantId))
+                        .map(ids -> ids.remove(message.getCorrelationId()))
+                        .ifPresent(expiringPromise -> {
+                            span.log("received command response");
+                            LOGGER.debug("received command response and cancelled timer");
+                            expiringPromise.tryCompleteAndCancelTimer(Future.succeededFuture(message));
+                        });
+            }
+        };
+    }
+
+    private Future<Void> subscribeForCommandResponse(final String tenantId, final Span span) {
+        final Handler<DownstreamMessage<KafkaMessageContext>> messageHandler = getCommandResponseHandler(tenantId, span);
+
+        if (commandResponseSubscriptions.get(tenantId) != null) {
+            span.log("command response consumer already exists");
+            LOGGER.debug("command response consumer already exists for tenant[{}]", tenantId);
+            return Future.succeededFuture();
```

Comment:
In this case here the created `messageHandler` with the span parameter doesn't get used.
That means that in this case a previously created `messageHandler` with a different span is used. That span is then the wrong one here.

So it looks to me the span parameter should be removed from `getCommandResponseHandler()`.

Reply:
I updated and pushed the changes related to the comments.